### PR TITLE
SCI Firewall Tests for macOS

### PR DESCRIPTION
**Pull request by vl-federico**
Provide utility functions for consistent JSON reporting of test results and errors. Note that the `yq(1)` is required to run the tests.

Please review the test names, display names, descriptions, scores and risk scores thoroughly ensuring consistency with the Windows version!

Here is a sample successful output:
```
[
[
  {
    "Name": "mFWEnabled",
    "DisplayName": "macOS Firewall enabled",
    "Description": "Checks whether the macOS firewall is enabled.",
    "Score": 10,
    "RiskScore": 100,
    "ResultData": {
      "Enabled": true
    }
  },
  {
    "Name": "mFWBlockAll",
    "DisplayName": "macOS Firewall block-all rule active",
    "Description": "Checks whether the block-all rule of the macOS firewall is active.",
    "Score": 3,
    "RiskScore": 70,
    "ResultData": {
      "Enabled": false
    }
  },
  {
    "Name": "mFWStealthMode",
    "DisplayName": "macOS Firewall stealth mode active",
    "Description": "Checks whether the stealth mode of the macOS firewall is active.",
    "Score": 2,
    "RiskScore": 80,
    "ResultData": {
      "Enabled": false
    }
  },
  {
    "Name": "mFWApprovedApps",
    "DisplayName": "macOS Firewall approved applications",
    "Description": "Provides a list of applications that may accept incoming connections.",
    "Score": 10,
    "RiskScore": 0,
    "ResultData": "{\"ApprovedApplications\":[\"/Applications/Spotify.app\",\"/System/Library/CoreServices/ControlCenter.app\",\"/Applications/Focusrite Control.app/Contents/Library/LoginItems/FocusriteControlServer.app\",\"/Applications/Telephone.app\",\"/Applications/Logic Pro X.app\",\"/Applications/zoom.us.app/Contents/Frameworks/airhost.app\",\"/System/Library/CoreServices/RemoteManagement/ARDAgent.app\"]}"
  }
]
```

Here is a sample output containing errors:
```
[
  {
    "Name": "mFWEnabled",
    "DisplayName": "macOS Firewall enabled",
    "Description": "Checks whether the macOS firewall is enabled.",
    "ErrorCode": 127,
    "ErrorMessage": "zsh_run_command:4: no such file or directory: /usr/libexec/ApplicationFirewall/socketf2ilterfw"
  },
  {
    "Name": "mFWBlockAll",
    "DisplayName": "macOS Firewall block-all rule active",
    "Description": "Checks whether the block-all rule of the macOS firewall is active.",
    "ErrorCode": 127,
    "ErrorMessage": "zsh_run_command:4: no such file or directory: /usr/libexec/ApplicationFirewall/socketf2ilterfw"
  },
  {
    "Name": "mFWStealthMode",
    "DisplayName": "macOS Firewall stealth mode active",
    "Description": "Checks whether the stealth mode of the macOS firewall is active.",
    "ErrorCode": 127,
    "ErrorMessage": "zsh_run_command:4: no such file or directory: /usr/libexec/ApplicationFirewall/socket2filterfw"
  },
  {
    "Name": "mFWApprovedApps",
    "DisplayName": "macOS Firewall approved applications",
    "Description": "Provides a list of applications that may accept incoming connections.",
    "ErrorCode": 127,
    "ErrorMessage": "zsh_run_command:4: no such file or directory: /usr/libexec/ApplicationFirewall/sockdetifilterfw"
  }
]
```

### CODE REVIEW

File: config/Security inventory/macOS/Firewall/Firewall.zsh [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    /usr/libexec/ApplicationFirewall/socketfilterfw --getglobalstate
+}
+
+vlCheckIsFirewallBlockallRuleEnabled()
+{
+  local riskScore=70
+  local expectedOutput="enabled"
+
+  vlCheckIsFeatureEnabledFromCommandOutput \
+    "FWBlockAll" \
```
[the remark this answers is not in the document]
Comment (vl-federico) [reply config/Security inventory/macOS/Firewall/Firewall.zsh:25]: That was a copy+paste error.

File: config/Security inventory/macOS/Firewall/Firewall.zsh [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+#
+
+vlCheckIsFirewallEnabled()
+{
+  local riskScore=100
+  local expectedOutput="enabled"
+
+  vlCheckIsFeatureEnabledFromCommandOutput \
+    "FWEnabled" \
+    "Firewall enabled" \
```
[the remark this answers is not in the document]
Comment (vl-federico) [reply config/Security inventory/macOS/Firewall/Firewall.zsh]: I added a prefix to both the test name and the display name for all tests on b739286369e95133d6eace2bffd93b1e6ddbcd93. Let me know if that works for you!

[the remark this answers is not in the document]
Comment (vl-federico) [reply config/Security inventory/macOS/Firewall/Firewall.zsh]: I added a prefix to both the test name and the display name for all tests on b739286369e95133d6eace2bffd93b1e6ddbcd93. Let me know if that works for you!

File: config/Security inventory/macOS/Firewall/Firewall.zsh [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+# Security and Compliance Inventory: Firewall Tests
+#
+
+vlCheckIsFirewallEnabled()
+{
+  local riskScore=100
+  local expectedOutput="enabled"
+
+  vlCheckIsFeatureEnabledFromCommandOutput \
+    "FWEnabled" \
```
[the remark this answers is not in the document]
Comment (vl-federico) [reply config/Security inventory/macOS/Firewall/Firewall.zsh]: See f40069c176c90cf8d6d1580792449b98ad14e9ab.

File: config/Security inventory/macOS/Firewall/Firewall.zsh (hunk 2 above)
[the remark this answers is not in the document]
Comment (vl-federico) [reply config/Security inventory/macOS/Firewall/Firewall.zsh]: As discussed on gchat, I removed the name prefixes in 4acbbeefa621ae8477bf57f7e9a2ffade763e2ab.

File: config/Security inventory/macOS/Firewall/Firewall.zsh (hunk 3 above)
[the remark this answers is not in the document]
Comment (vl-federico) [reply config/Security inventory/macOS/Firewall/Firewall.zsh]: As discussed on gchat, I removed the name prefixes in 4acbbeefa621ae8477bf57f7e9a2ffade763e2ab.

File: config/Security inventory/macOS/Firewall/Firewall.zsh [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    "Checks whether the macOS firewall is enabled." \
+    "$riskScore" \
+    "$expectedOutput" \
+    /usr/libexec/ApplicationFirewall/socketfilterfw --getglobalstate
+}
+
+vlCheckIsFirewallBlockallRuleEnabled()
+{
+  local riskScore=70
+  local expectedOutput="enabled"
```
Comment (martin-kretzschmar) [inline config/Security inventory/macOS/Firewall/Firewall.zsh:22]: `enabled` seems to be the wrong expected result to identify the state.

File: config/Security inventory/macOS/Firewall/Firewall.zsh [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+vlGetFirewallApprovedApps()
+{
+  local testName="FWApprovedApps"
+  local testDisplayName="macOS Firewall approved applications"
+  local testDescription="Provides a list of applications that may accept incoming connections."
+  # This test is only informational and always returns a fixed test score
+  local testScore=10
+  local riskScore=0
+
+  vlRunCommand /usr/libexec/ApplicationFirewall/socketfilterfw --listapps
```
Comment (martin-kretzschmar) [inline config/Security inventory/macOS/Firewall/Firewall.zsh:56]: In my test `/usr/libexec/ApplicationFirewall/socketfilterfw --listapps` returns:

```ALF: total number of apps = 6

1 :  /System/Library/CoreServices/RemoteManagement/ARDAgent.app
 	 ( Allow incoming connections )

2 :  /System/Library/CoreServices/ControlCenter.app
 	 ( Allow incoming connections )

3 :  /opt/homebrew/Cellar/iperf3/3.15/bin/iperf3
 	 ( Allow incoming connections )

4 :  /opt/homebrew/Cellar/netcat/0.7.1/bin/netcat
 	 ( Allow incoming connections )

5 :  /Applications/Elgato Stream Deck.app
 	 ( Allow incoming connections )

6 :  /Library/Application Support/Logitech.localized/LogiOptionsPlus/logioptionsplus_agent.app
 	 ( Allow incoming connections )
```

While the script output shows:

```  {
    "Name": "FWApprovedApps",
    "DisplayName": "macOS Firewall approved applications",
    "Description": "Provides a list of applications that may accept incoming connections.",
    "Score": 10,
    "RiskScore": 0,
    "ResultData": "{\"ApprovedApplications\":[\"/System/Library/CoreServices/RemoteManagement/ARDAgent.app\",\"/System/Library/CoreServices/ControlCenter.app\",\"/opt/homebrew/Cellar/iperf3/3.15/bin/iperf3\",\"/opt/homebrew/Cellar/netcat/0.7.1/bin/netca\",\"/Applications/Elgato Stream Deck.app\",\"/Library/Application Support/Logitech.localized/LogiOptionsPlus/logioptionsplus_agent.app\"]}"
  }
```
```suggestion
  vlRunCommand /usr/libexec/ApplicationFirewall/socketfilterfw --listapps
```

Notice the missing "t" in `netca`.

**Review by martin-kretzschmar — CHANGES_REQUESTED** [review]

File: config/Security inventory/macOS/Firewall/Firewall.zsh (hunk 4 above)
Comment (vl-federico) [reply config/Security inventory/macOS/Firewall/Firewall.zsh]: Good catch! See 5f5f00020082c29842c3ff983bf1d21d277a7f99.

File: config/Security inventory/macOS/Firewall/Firewall.zsh (hunk 5 above)
Comment (vl-federico) [reply config/Security inventory/macOS/Firewall/Firewall.zsh:56]: Good catch! Fixed in daa2c1f648b2355cc36a5f5583b5ab737e2f3831

**Review by martin-kretzschmar — APPROVED** [review]

**Pull request merged**